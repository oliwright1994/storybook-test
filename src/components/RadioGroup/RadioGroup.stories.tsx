import createStory from '@bright/storybook/createStory'
import { boolean, text } from '@storybook/addon-knobs'
import React from 'react'

import RadioGroup from '.'

createStory('COMPONENTS | RadioGroup', [
  {
    label: 'RadioGroup',
    jsx: () => {
      let errorMessage
      const description = text('Description', 'This is a RadioGroup with three options')
      const isInline = boolean('Inline', false)
      const isRequired = boolean('Required', false)
      const isInvalid = boolean('Invalid', false)
      if (isInvalid) {
        errorMessage = text('Error Message', 'There has been an error')
      }

      return (
        <RadioGroup
          defaultValue="1"
          errorMessage={errorMessage}
          description={description}
          isRequired={isRequired}
          isInline={isInline}
        >
          <RadioGroup.Radio id="1" value="1">
            Value 1
          </RadioGroup.Radio>
          <RadioGroup.Radio id="2" value="2">
            Value 2
          </RadioGroup.Radio>
          <RadioGroup.Radio id="3" value="3">
            Value 3
          </RadioGroup.Radio>
        </RadioGroup>
      )
    },
  },
])
