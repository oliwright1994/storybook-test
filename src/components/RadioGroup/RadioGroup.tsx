import Radio from '@bright/components/RadioGroup/Radio'
import Stack from '@bright/components/Stack'
import { Space } from '@bright/constants'
import {
  FormControl,
  FormErrorMessage,
  FormHelperText,
  IRadioGroup as ChakraIRadioGroup,
  RadioGroup as ChakraRadioGroup,
} from '@chakra-ui/core'
import React, { FC } from 'react'

interface IRadioGroupProps extends Pick<ChakraIRadioGroup, 'onChange' | 'value'> {
  errorMessage?: string
  description?: string
  defaultValue?: string | number
  isRequired?: boolean
  isInline?: boolean
  isInvalid?: boolean
}

interface IRadioGroup extends FC<IRadioGroupProps> {
  Radio: typeof Radio
}

const RadioGroup: IRadioGroup = ({
  children,
  description,
  onChange,
  isInvalid,
  isRequired,
  defaultValue,
  isInline,
  value,
  errorMessage,
}) => {
  return (
    <FormControl isInvalid={isInvalid} isRequired={isRequired}>
      <Stack space={Space.xxs}>
        <ChakraRadioGroup
          value={value}
          onChange={onChange}
          isInline={isInline}
          defaultValue={defaultValue}
          alignItems="baseline"
        >
          {children}
        </ChakraRadioGroup>
        {description && <FormHelperText margin={0}>{description}</FormHelperText>}
        {errorMessage && <FormErrorMessage>{errorMessage}</FormErrorMessage>}
      </Stack>
    </FormControl>
  )
}

RadioGroup.Radio = Radio

export default RadioGroup
