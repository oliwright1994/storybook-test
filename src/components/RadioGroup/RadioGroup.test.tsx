import { fireEvent } from '@testing-library/react'
import React from 'react'
import { renderWithBrightTheme } from 'testing/helpers/themeHelper'

import RadioGroup from '.'

describe('<RadioGroup />', () => {
  const defaultGroupProps = {
    description: 'This is a description',
    errorMessage: 'This is an error message',
  }
  const getLabel = (id: string): string => `Option ${id}`
  const radioButtons = ['1', '2'].map(id => (
    <RadioGroup.Radio key={id} id={id} value={id}>
      {getLabel(id)}
    </RadioGroup.Radio>
  ))

  it('should render as expected', () => {
    const { container } = renderWithBrightTheme(<RadioGroup {...defaultGroupProps}>{radioButtons}</RadioGroup>)

    expect(container).toMatchSnapshot()
  })

  it('should render child radio inputs as expected', () => {
    const { container } = renderWithBrightTheme(<RadioGroup {...defaultGroupProps}>{radioButtons}</RadioGroup>)

    expect(container.querySelectorAll('input[type="radio"]')).toHaveLength(2)
    expect(container).toHaveTextContent(getLabel('1'))
    expect(container).toHaveTextContent(getLabel('2'))
  })

  it('should render description', () => {
    const { container } = renderWithBrightTheme(<RadioGroup {...defaultGroupProps}>{radioButtons}</RadioGroup>)

    expect(container).toHaveTextContent(defaultGroupProps.description)
  })

  it('should render error message when isInvalid', () => {
    const { container } = renderWithBrightTheme(
      <RadioGroup {...defaultGroupProps} isInvalid>
        {radioButtons}
      </RadioGroup>
    )

    expect(container).toHaveTextContent(defaultGroupProps.errorMessage)
  })

  it('should fire onChange function', () => {
    const handleChange = jest.fn()
    const { getByText } = renderWithBrightTheme(
      <RadioGroup {...defaultGroupProps} onChange={handleChange}>
        {radioButtons}
      </RadioGroup>
    )
    const radioInput = getByText(getLabel('2'))

    expect(handleChange).toHaveBeenCalledTimes(0)
    fireEvent.click(radioInput)
    expect(handleChange).toHaveBeenCalledTimes(1)
  })
})
