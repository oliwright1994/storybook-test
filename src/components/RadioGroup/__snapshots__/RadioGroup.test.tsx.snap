// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RadioGroup /> should render as expected 1`] = `
<div>
  <div
    class="css-0"
    role="group"
  >
    <div
      class="css-0"
    >
      <li
        class="css-1lxwves"
      >
        <div
          class="css-1y3jl3a"
          role="radiogroup"
        >
          <div
            class="css-1x9mqzb"
          >
            <label
              class="css-jwmyb8"
              for="1"
            >
              <input
                class="css-1gy22ay"
                id="1"
                type="radio"
                value="1"
              />
              <div
                aria-hidden="true"
                class="css-1e2c42x"
                type="radio"
              >
                <span
                  class="css-z7lnx4"
                />
              </div>
              <div
                class="css-6qsuox"
              >
                Option 1
              </div>
            </label>
          </div>
          <div
            class="css-13o7eu2"
          >
            <label
              class="css-jwmyb8"
              for="2"
            >
              <input
                class="css-1gy22ay"
                id="2"
                type="radio"
                value="2"
              />
              <div
                aria-hidden="true"
                class="css-1e2c42x"
                type="radio"
              >
                <span
                  class="css-z7lnx4"
                />
              </div>
              <div
                class="css-6qsuox"
              >
                Option 2
              </div>
            </label>
          </div>
        </div>
      </li>
      <li
        class="css-qzdso"
      >
        <p
          class="css-1kvhzt2"
        >
          This is a description
        </p>
      </li>
      <li
        class="css-qzdso"
      />
    </div>
  </div>
</div>
`;
